### ISSUE TITLE
How to use agora in Ionic 4

### ISSUE THREAD

**Issue by abhibly** [AgoraIO-Community/Angular-Agora-RTC#15]

**Comment by sureshtpix** [comment]
Did you find any solution to this?

**Comment by cavalryjim** [comment]
I don't think the angular plug-in will work in the Ionic framework without some additional effort to access the camera and microphone.  I would be willing the partner in a solution if anyone else is interested.

**Comment by marianoir** [comment]
You have to allow the access to the camera, record_audio and modify_audio_settings. Try to add that manually on androidmanifest. 

Or try , on the phone,  go to configuration --> aplications and give manually access to the camera and microphone to the app. 

The problem is that the "publish" method on the android phone is not receiving the "publish stream has been published" message so i think that something is missing.

**Comment by KirankumarDafda** [comment]
I am also working on this for IONIC 4 but not able to open camera, now trying to set it manually or will add camera plugin into this

**Comment by innovativethinkingptyltd** [comment]
@KirankumarDafda did you manage with this?

**Comment by rafaelfranco** [comment]
> You have to allow the access to the camera, record_audio and modify_audio_settings. Try to add that manually on androidmanifest.
> 
> Or try , on the phone, go to configuration --> aplications and give manually access to the camera and microphone to the app.
> 
> The problem is that the "publish" method on the android phone is not receiving the "publish stream has been published" message so i think that something is missing.

Did you find any solution to this?

**Comment by samalfaro92** [comment]
Did someone manage to fix this?

**Comment by plutoless** [comment]
for people who are having issues with this platform. is there any chance you can collect the sdk logs for me so that i can analyze what the issue is?

**Comment by yohan-pannet-bbo** [comment]
I'm not using the exact same plugin (ngx-agora) but i had a similar issue. Hope this will help:

As @cavalryjim suggested, this is indeed a permission issue.

I added permissions declaration in the config.xml:
`<config-file parent="/manifest" target="AndroidManifest.xml" xmlns:android="http://schemas.android.com/apk/res/android">
        <uses-permission android:name="android.permission.CAMERA" />
        <uses-feature android:name="android.hardware.camera" android:required="true" />
        <uses-feature android:name="android.hardware.microphone" android:required="true" />
        <uses-permission android:name="android.permission.RECORD_AUDIO" />
        <uses-permission android:name="android.permission.MODIFY_AUDIO_SETTINGS" />
    </config-file>`

And i used [cordova-plugin-android-permissions](https://www.npmjs.com/package/cordova-plugin-android-permissions) to manage permission requests.

I check permission (permissions.checkPermission(permission, successCallback, errorCallback)) for permissions.CAMERA & permissions.RECORD_AUDIO
And then I request permissions (permissions.requestPermissions(permissions, successCallback, errorCallback)) for permissions not granted yet.

rq: be carefull not to call requestPermissions with an empty list because it returns an error. 

Regards

**Comment by RosanneJiaJia** [comment]
Did anyone manage to get this working in ionic native device? (none browser version). Asking because I've got error like "Agora-SDK [ERROR]: Web RTC is not supported in this browser“， just wondering whether it is possible to make it working.

**Comment by yohan-pannet-bbo** [comment]
@RosanneJiaJia, what device and os version are you testing on?
Web RTC wasn't well supported on ios before 14.5.
I wasn't able to publish stream in ios before this version...
On android side, appart from permissions (see above), i didn't have issue to publish and subscribe.

**Comment by RosanneJiaJia** [comment]
@yohan-pannet-bbo, thank you for your input. I am testing in Iphone with ios 14.5.1

After I upgrade to IOS 14.5, I still couldn't publish the stream. Did you manage to get it working in IOS 14.5?

Error log:
2021-05-09 16:24:48.441176+0100 ERROR: 16:24:48:440 Agora-SDK [ERROR]: [F70B2]**None Ice Candidate not allowed**
2021-05-09 16:24:59.101236+0100 ERROR: 16:24:59:99 Agora-SDK [ERROR]: Request Timeout 10 [object Object]
2021-05-09 16:24:59.101478+0100 ERROR: 16:24:59:100 Agora-SDK [ERROR]: Publish Failed: TIMEOUT


Regards

**Comment by plutoless** [comment]
@RosanneJiaJia can you make it working in safari browser? in same network?

**Comment by RosanneJiaJia** [comment]
@plutoless yes, browser (safari, chrome, firefox) version is working in native device. I just can't get it working natively inside the IOS app. According to @yohan-pannet-bbo, it seems to be possible to get this plugin working natively. But I haven't succeeded so far.

**Comment by plutoless** [comment]
@RosanneJiaJia for regular versions of iOS it's not possible (i.e. webrtc is not fully suppprted). For latest version it is said to be possible but test is not yet enough on it. Currently webrtc support on iOS WebView is still very poor it's not recommended to use this as a production platform.

**Comment by yohan-pannet-bbo** [comment]
@plutoless, i've managed to publish my local stream on ios app on 14.5 (doesn't work on 14.4). But since it's very fresh, maybe there is still instabilities... for instance i still have some bugs (see bellow).

@RosanneJiaJia, out of curiosity, what do you use to build your native? ionic? phonegap? 
You say you can't publish your stream from the app, can it receive/subscribe to a remote one?
Have you tried on android? or maybe it is not a targeted platform.

I still have a remaining bug: on ipad ios 14.5 or higher (and not on iphone where there is no bug), i can publish my local stream but cannot receive incoming ones. It also trigger the "None Ice Candidate not allowed"

Have you looked at the agora's page about "None Ice Candidate not allowed":
https://docs.agora.io/en/Agora%20Platform/faq/console_error_web

**Comment by atishabhang** [comment]
@abhibly I managed to have native screens for both audio and video calls in android.
All working perfectly. But i don't know how to make this work in ios.  I have no idea how to present native screens I followed the capacitor docs to present native screens but no success. 
If anyone managed to run native screens for audio or video with the agora SDK in iOS please share.
If anyone wants to know the android working i'll be happy to share.

**Comment by kaushik657** [comment]
I implemented agora web sdk for android. It's working fine.
But in case of iOS not working, only remote view is showing when join channel from android means I received stream but unable to send because local media not attached. I am using without webrtc plugin, when I install webrtc plugin nothing worked neither local view nor remote view.
